### ISSUE TITLE
Question about Figure 3

### ISSUE THREAD

**Issue by ybdai7** [megvii-research/FSSD_OoD_Detection#10]
Hi. I am wondering how to get the initial concentration as it is shown in Figure 3 in the paper. Could the default pytorch initialization method for ResNet 34 get the same result?

I am trying to reproduce the concentration result using ResNet-34 for CIFAR-10. While the penultimate feature of OoD data (uniformly generated data ranging form [0,1]) seems to concentrate to a fix point, the feature representation of CIFAR-10 images got from penultimate layer seems quite sparse, as you can see in the following figure where 10 stands for OoD data.

Could you kindly provide the code generating Figure 3 in the paper? or please enlighten me if there is anything i am missing.

![Figure_1](https://github.com/megvii-research/FSSD_OoD_Detection/assets/95686360/177b3f26-829e-4ca0-87b4-7fee6fdccbe1)

**Comment by andrehuang** [comment]
Hi Yanbo,

Thanks for your interest in our paper. Unfortunately, the code of the plot generation is no longer available. But I can share some knowledge from my memory.

Yes you can plot the figure using a pretrained Pytorch ResNet34. If I remember correctly, we were using this codebase to train them: https://github.com/kuangliu/pytorch-cifar. But we also tested with many pretrained models. 

For the plotting, are you using PCA? If you use tSNE, you won't be able to plot it as in Figure 3. Similar figures (without the concentration point) has been shown in many other papers as well. Hope that helps.

Best,
Haiwen

**Comment by ybdai7** [comment]
Hi Haiwen,

Thanks so much for the prompt reply. Yes, i am using tSNE and i guess thats the reason. 

Also, i am wondering what's the reason behind this concentration effect at the very initialization stage of neural network. As i am new to this field (i mainly work on backdoors in FL and i find the concentration effect very interesting), could you kindly provide me with maybe some papers illustrating this effect or some related?

Thanks again for your prompt response which helps a lot. And also thanks for the great paper which inspires me a lot.

Thanks,
Yanbo

**Comment by ybdai7** [comment]
Hi Haiwen,

And i am curious about the specific meaning of the "similar figures" you said. Do you mean the difference between pca and tsne shown in other papers?

Thanks,
Yanbo

**Comment by andrehuang** [comment]
Hi Yanbo,

Our paper is afaik the first to illustrate and explain this concentration effect. To understand this theoretically, you can check out the explanation in our paper (Equations (2)-(4)). 

I can also give you a more intuitive understanding: The training process of neural networks is basically making the neural features more activated to the patterns in the data. So as training goes on, the features will be dragged away from the concentration point. At the same time, the NNs will be activated more and more by your data. 

By "similar figures", I mean the PCA figure with rays of logits shooting from a center. Maybe you have also seen such figures before. Unfortunately, I don't have the names of those papers. But I believe there are plenty of them.

Best,
Haiwen

**Comment by ybdai7** [comment]
Thanks. That surely helps a lot. Have a good day.
